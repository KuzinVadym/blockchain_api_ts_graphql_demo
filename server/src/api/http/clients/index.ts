import { IClients } from './shared/interfaces';
import { IGetStateFunction } from '../../../shared/interfaces/IGetStateFunction';
import { createBlockchainHttpClient } from './blockchain/BlockchainClient';

export interface ICreateHttpClients {
  (getState: IGetStateFunction): IClients;
}

export function createHttpClients(getState: IGetStateFunction): IClients {
  return {
    blockchain: createBlockchainHttpClient(getState)
  };
}
