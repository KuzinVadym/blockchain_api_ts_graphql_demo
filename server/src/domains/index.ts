import { IServices } from './shared/interfaces/IServices';
import { IGetStateFunction } from '../shared/interfaces/IGetStateFunction';
import { createBlockchainHttpService } from './blockchain/services/http';

export interface ICreateHttpServices {
  (getState: IGetStateFunction): IServices;
}

export function createHttpServices(getState: IGetStateFunction): IServices {
  return {
    blockchain: createBlockchainHttpService(getState)
  };
}
