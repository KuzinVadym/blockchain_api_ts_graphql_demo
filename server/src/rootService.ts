import 'reflect-metadata';
import http, { Server } from 'http';
import express from 'express';
import { ApolloServer } from 'apollo-server-express';

import { getSchema } from './api/http/graphql/schema';
import { IAppServer } from './shared/interfaces/IAppServer';
import { ILogger, IState } from './shared/interfaces/IState';
import { ISettings } from './shared/interfaces/ISettings';
import { IServices } from './domains/shared/interfaces/IServices';
import { IClients } from './shared/interfaces/IClients';
import { ICreateHttpServices } from './domains';
import { ICreateHttpClients } from './api/http/clients';

export class AppServer implements IAppServer {
  private server: ApolloServer;
  private httpServer: Server;
  private logger: ILogger;
  private settings: ISettings;
  private httpServices: IServices;
  private httpClients: IClients;

  constructor(settings: ISettings, logger: ILogger) {
    this.settings = settings;
    this.logger = logger;
  }

  public getState(): IState {
    return {
      logger: this.logger,
      httpServices: this.httpServices,
      httpClients: this.httpClients
    };
  }

  async withHttpClients(createHttpClients: ICreateHttpClients) {
    const getState = () => {
      return this.getState();
    };
    this.httpClients = createHttpClients(getState);
  }

  async withHttpServices(createHttpServices: ICreateHttpServices) {
    const getState = () => {
      return this.getState();
    };
    this.httpServices = createHttpServices(getState);
  }

  async init(): Promise<void> {
    this.logger.info('Init Apollo Server');

    const getState = () => {
      return this.getState();
    };

    const app = express();

    const schema = await getSchema();

    this.server = new ApolloServer({
      schema,
      context: (context) => {
        return { getState };
      }
    });

    this.server.applyMiddleware({ app });

    this.httpServer = http.createServer(app);

    // server.installSubscriptionHandlers(this.httpServer);
  }

  async initTest(): Promise<ApolloServer> {
    this.logger.info('Init Apollo Server for tests');

    const getState = () => {
      return this.getState();
    };

    const schema = await getSchema();

    return new ApolloServer({
      schema,
      context: (context) => {
        return { getState };
      }
    });
  }

  listen(): void {
    this.httpServer.listen(this.settings.port, () => {
      this.logger.info(
        `ðŸš€ Server ready at http://localhost:${this.settings.port}${this.server.graphqlPath}`
      );
    });
  }
}
