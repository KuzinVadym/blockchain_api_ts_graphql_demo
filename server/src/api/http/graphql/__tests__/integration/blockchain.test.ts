import { ApolloServer } from 'apollo-server-express';
import {
  ApolloServerTestClient,
  createTestClient
} from 'apollo-server-testing';
import pino from 'pino';
import { AppServer } from '../../../../../rootService';
import { createHttpServices } from '../../../../../domains';
import { createHttpClients } from '../../../clients';

const logger = pino();

describe('blockchain integration tests', () => {
  let server: ApolloServer;

  beforeAll(async () => {
    const testAppSrv = new AppServer({ port: '4000' }, logger);
    await testAppSrv.withHttpClients(createHttpClients);
    await testAppSrv.withHttpServices(createHttpServices);
    server = await testAppSrv.initTest();
  });

  it('finds block by hash', async () => {
    const testQuery = `
    query BlockChainQuery($hash: String!){
      rawBlock(hash: $hash) {
        size
        block_index
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    const result = await query({
      query: testQuery,
      variables: {
        hash: '000000000000000000082bac0859c43f89c35423049712658ee9a0963c15aaf3'
      }
    });

    expect(result.data.rawBlock).toMatchSnapshot();
  });

  it('returns blocks data', async () => {
    const testQuery = `{
      blocks {
        hash
        time
        height
      }
    }
  `;

    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: {} });
    expect(res.data.blocks.length).toBeGreaterThan(0);
  });
});
