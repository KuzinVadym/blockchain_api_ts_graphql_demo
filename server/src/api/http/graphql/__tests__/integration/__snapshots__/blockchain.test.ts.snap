// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blockchain integration tests finds block by hash 1`] = `
Object {
  "block_index": 0,
  "size": 1819006,
}
`;
